lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,628 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! A module for paying Lightning invoices.
+//!
+//! Defines an [`InvoicePayer`] utility for paying invoices, parameterized by [`Payer`] and
+//! [`Router`] traits. Implementations of [`Payer`] provide the payer's node id, channels, and means
+//! to send a payment over a [`Route`]. Implementations of [`Router`] find a [`Route`] between payer
+//! and payee using information provided by the payer and from the payee's [`Invoice`].
+//!
+//! [`InvoicePayer`] caches each [`Invoice`] by its `payment_hash` so that it can retry the payment
+//! if it fails. It accomplishes this by implementing [`EventHandler`] which decorates a
+//! user-provided handler. It will intercept any [`Event::PaymentPathFailed`] events and retry the
+//! payment a fixed number of times before failing or succeeding as appropriate.
```

Comment:
Do we guarantee that we'll only pass along a `PaymentPathFailed` event if `all_paths_failed`? Or might some paths still be pending when we pass along the event? Could be worth documenting

Reply:
Reworded based offline discussion.